### PR TITLE
io safety: update Unix explanation to use `Arc`

### LINKED ISSUE #124384: Module docs for std::os::unix::io inconsistent with I/O Safety discussion in std::io
### Location

https://doc.rust-lang.org/stable/std/os/unix/io/index.html

### Summary

In std::io, [it correctly says](https://doc.rust-lang.org/stable/std/io/index.html#io-safety):
>Note that exclusive ownership of a file descriptor does not imply exclusive ownership of the underlying kernel object that the file descriptor references (also called “file description” on some operating systems). File descriptors basically work like [Arc](https://doc.rust-lang.org/stable/std/sync/struct.Arc.html): when you receive an owned file descriptor, you cannot know whether there are any other file descriptors that reference the same kernel object. However, when you create a new kernel object, you know that you are holding the only reference to it. Just be careful not to lend it to anyone, since they can obtain a clone and then you can no longer know what the reference count is! In that sense, [OwnedFd](https://doc.rust-lang.org/stable/std/os/fd/struct.OwnedFd.html) is like Arc and [BorrowedFd<'a>](https://doc.rust-lang.org/stable/std/os/fd/struct.BorrowedFd.html) is like &'a Arc

However the table in std::io::unix::io has the table saying:
> [BorrowedFd<'a>](https://doc.rust-lang.org/stable/std/os/fd/struct.BorrowedFd.html)	=> &'a _
[OwnedFd](https://doc.rust-lang.org/stable/std/os/fd/struct.OwnedFd.html)	=> Box<_>

This is incorrect and potentially very misleading, because the result of `dup`ping a file descriptor creates a new fd which shares the same underlying file structure in the kernel. In particular, for seekable fds they share the current file offset (ie, used by `read` `write` and `lseek`) and in some cases, file locking state. (In addition to inode state and underlying file/pipe/etc contents.)

The language here should reflect the language in std::io.

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/124384

Cc @jsgf

### CODE REVIEW

File: library/std/src/os/unix/io/mod.rs [hunk 1]
```diff
@@ -23,22 +23,33 @@
 //! either by adding `unsafe` to APIs that dereference `RawFd` values, or by
 //! using to `BorrowedFd` or `OwnedFd` instead.
 //!
+//! The use of `Arc` for borrowed/owned file descriptors may be surprising.
+//! Unix file descriptors are mere references to internal *file descriptions*,
```
Comment (jsgf) [inline library/std/src/os/unix/io/mod.rs:27]: Can't say I'm a huge fan of the term "file description" (it's too similar to "file descriptor"), but I guess we're stuck with it if that's what Posix uses.

**Comment by jsgf** [comment]
Yep, looks good to me.

File: library/std/src/os/unix/io/mod.rs (hunk 1 above)
Comment (ChrisDenton) [reply library/std/src/os/unix/io/mod.rs:27]: The full name is [open file description](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_258) which might be slightly less confusing? Maybe saying "kernel descriptor" or "kernel file descriptor" would help clarify. But sticking with the nomenclature used elsewhere does make sense.

Comment (RalfJung) [reply library/std/src/os/unix/io/mod.rs]: I have added the "open".

**Pull request merged**